File: 11/ruby/array_sum.rb
# array_sum.rb sums the values in an array
#
# Begun by: Prof. Adams, for CS 214 at Calvin College.
# Completed by:
# Date:
#
# Usage: ruby array_sum.rb <inFileName>
######################################################

# make certain the input file was specified on the command-line
def checkCommandLine
  if ARGV.length != 1
     puts "\nUsage: sum <fileName>\n\n"
     exit(1)
  end
end

# return: an array containing the values from the file
#  whose name was specified on the command-line
def readFile
   strArray = IO.readlines( ARGV[0] )
   intArray = Array.new( Integer(strArray[0]) )
   for i in 2..strArray.size
      if  !strArray[i].nil?
         intArray[i-2] = Integer( strArray[i].chomp )
      end
   end
   intArray 
end

# Return: the sum of the values 
def sum(values)
   result = 0
   values.each{ |number| result += number }
   result
end

def main
   checkCommandLine
   values = readFile
   startTime = Time.now
   total = sum(values)
   endTime = Time.now
   interval = (endTime - startTime).to_f
   puts "\nThe sum is #{total}\n"
   printf(" and computing it took %.9f seconds.\n\n", interval)
end

main

File: 11/c++/array_sum.cpp
/* array_sum.cpp sums the values in a file
 *  by reading them into an array and summing them there.
 *
 * Begun by: Prof. Adams, for CS 214 at Calvin College.
 * Completed by:
 * Date:
 *
 * Usage: array_sum <inFileName>
 */

#include <iostream>     // cin, cout, etc
#include <fstream>      // ifstream, ofstream
#include <vector>       // vector
#include <cassert>      // assert()
#include <cstdlib>      // exit()
#include "NanoTimer.h"  // nanosecond timer
#include <iomanip>      // formatted output
using namespace std;

void checkCommandLine(int argc);
void readFile(const string& fileName, vector<int>& v);
long sum(const vector<int>& v);
void display(long total, const NanoTimer& nTimer);

int main(int argc, char** argv) {
   checkCommandLine(argc);
   vector<int> values;
   readFile( argv[1], values );
   long total;
   NanoTimer nTimer;
   nTimer.start();
   total = sum(values);
   nTimer.stop();
   display(total, nTimer);
}

void checkCommandLine(int argc) {
   if (argc != 2) {
      cerr << "\nUsage: array_sum <fileName>\n\n";
      exit(1);
   }
}

long sum(const vector<int>& v) {
   long result = 0;

   for (unsigned i = 0; i < v.size(); i++) {
      result += v[i];
   }

   return result;
}


void readFile(const string& fileName, vector<int>& v) {
   ifstream fin( fileName.c_str() );
   assert( fin.is_open() );

   int number;
   fin >> number;
   v.reserve(number);
   while (true) {
      fin >> number;
      if ( fin.eof() ) break;
      v.push_back(number);
   }
   
   fin.close();
}

void display(long total , const NanoTimer& nTimer) {
   cout << "\nThe sum is " << total 
        << "\n and computing it took " 
        << fixed << showpoint << setprecision(9) 
        << nTimer 
        << " secs.\n" << endl;

}
